specs/ContextMenuRequested.md

```diff
@@ -0,0 +1,764 @@
+                                {
+                                    MessageBox(
+                                        m_appWindow->GetMainWindow(), pageString.c_str(),
+                                        L"Display Page Uri", MB_OK);
+                                });
+                                return S_OK;
+                            })
+                            .Get(),
+                        &m_customItemSelectedToken);
+                    CHECK_FAILURE(items->AddValueAtIndex(itemsCount, newMenuItem.get()));
+                }
+                return S_OK;
+            })
+            .Get(),
+        &m_contextMenuRequestedToken);
+```
+
+## Win32 C++ Use Data to Display Custom Context Menu
+
+The developer can use the data provided in the Event arguments to display a custom context menu with entries of their choice. For this case, the developer specifies Handled to be true and requests a deferral. Deferral of this event should be completed when the user selects a context menu item (either the app developer will handle the case, or can return the selected option to the WebView) or when the end user dismisses the context menu by clicking outside of the context menu for example.
+
+ ```cpp
+    webview2->add_ContextMenuRequested(
+        Callback<ICoreWebView2ContextMenuRequestedEventHandler>(
+            [this](
+                ICoreWebView2* sender,
+                ICoreWebView2ContextMenuRequestedEventArgs* args)
+            {
+                auto showMenu = [this, args]
+                {
+                    wil::com_ptr<ICoreWebView2ContextMenuItemCollection> items;
+                    CHECK_FAILURE(args->get_MenuItems(&items));
+                    CHECK_FAILURE(args->put_Handled(true));
+                    HMENU hPopupMenu = CreatePopupMenu();
+                    AddMenuItems(hPopupMenu, items);
+                    HWND hWnd;
+                    m_controller->get_ParentWindow(&hWnd);
+                    SetForegroundWindow(hWnd);
+                    wil::com_ptr<ICoreWebView2ContextMenuInfo> parameters;
+                    CHECK_FAILURE(args->get_ContextMenuInfo(&parameters));
+                    POINT locationInControlCoordinates;
+                    POINT locationInScreenCoordinates;
+                    CHECK_FAILURE(parameters->get_Location(&locationInControlCoordinates));
+                    // get_Location returns coordinates in relation to upper left Bounds of the WebView2.Controller. Will need to convert to Screen coordinates to display the popup menu in the correct location.
+                    ConvertToScreenCoordinates(locationInControlCoordinates, locationInScreenCoordinates);
+                    UINT32 selectedCommandId = TrackPopupMenu(hPopupMenu, TPM_TOPALIGN | TPM_LEFTALIGN | TPM_RETURNCMD, locationInScreenCoordinates.x, locationInScreenCoordinates.y, 0, hWnd, NULL);
+                    CHECK_FAILURE(args->put_SelectedCommandId(selectedCommandId));
```

Comment:
Is there a range of command IDs that CoreWebView2 promises not to use, so apps can add their additional menu items with those command IDs?

Or should the app just create a dummy CreateContextMenuItem, add it to the menu, and then read it back to see what CommandId the system assigned to that menu item (presumably a unique one)?

Reply:
Currently the second option is being done where the system assigns unique command IDs to those created by the developer (devs won't have to worry about adding their own CommandID).